client/data/agency-dashboard/use-sync-jetpack-connection-status.ts

```diff
@@ -0,0 +1,61 @@
+import { useQuery, useQueryClient } from '@tanstack/react-query';
+import { useContext, useEffect, useMemo } from 'react';
+import SitesOverviewContext from 'calypso/jetpack-cloud/sections/agency-dashboard/sites-overview/context';
+import { Site } from 'calypso/jetpack-cloud/sections/agency-dashboard/sites-overview/types';
+import wpcom from 'calypso/lib/wp';
+
+const useSyncJetpackConnectionStatus = (
+	isPartnerOAuthTokenLoaded: boolean,
+	isConnectionHealthy: boolean,
+	siteId: number
+) => {
+	const queryClient = useQueryClient();
+	const { search, currentPage, filter, sort } = useContext( SitesOverviewContext );
+
+	const dashboardSitesQueryKey = useMemo(
+		() => [ 'jetpack-agency-dashboard-sites', search, currentPage, filter, sort ],
+		[ currentPage, filter, search, sort ]
+	);
+
+	const { data } = useQuery( {
+		queryKey: [ 'jetpack-agency-test-connection', { siteId, isConnectionHealthy } ],
+		queryFn: () =>
+			wpcom.req.get(
+				{
+					path: `/jetpack-blogs/${ siteId }/test-connection`,
+					apiNamespace: 'rest/v1.1',
+				},
+				{
+					// We call the current health state "stale", as it might be different than the actual state.
+					is_stale_connection_healthy: Boolean( isConnectionHealthy ),
+				}
+			),
+		enabled: isPartnerOAuthTokenLoaded,
+		// We don't want to trigger another API request to the /test-connection endpoint for a given site
+		// one minute after the first successful one.
+		staleTime: 1000 * 60,
+	} );
+
+	useEffect( () => {
+		// We are setting the "is_connected" property of the site in the query cache of the "sites"
+		// API to the value returned by the /test-connection endpoint. We are doing this to filter
+		// the site with connection issues easily in the UI and to have a single source of truth for
+		// the connection state of a site to avoid inconsistencies.
+		queryClient.setQueryData( dashboardSitesQueryKey, ( oldSites: any ) => {
+			return {
+				...oldSites,
```

Comment:
What if the old sites data is no longer relevant because one or more of the sites have been disconnected in the meantime? With that in mind, It feels like we should be maintaining a fresh connection state every time, ignoring the previous `oldSites`. But then that should be achievable with a straightforward callback after the `wpcom.req.get()` call, similar to this [Service Layer Pattern](https://github.com/TanStack/query/discussions/5279#discussioncomment-5632938)?

Reply:
After taking some time to think this through I'd say there's a lot of value in changing how the query and the data are constructed, which then would align with what you're suggesting. For the sake of this PR (we're not changing any of the current functionality) though I'd follow up with another one 'cause it certainly requires a bit more effort and exceeds the purpose of this PR if that's okay with you. Alternatively I could close this one and do the rework entirely in a different PR.

Reply:
Thanks for considering it. Given that the suggestion here might introduce bugs related to stale connection data, so I'd recommend either refactoring it here or in another PR if that works for you.